### PR TITLE
Added key equality mode

### PR DESCRIPTION
**Pull request by carldebilly**

### CODE REVIEW

File: src/Uno.CodeGen/EqualityGenerator.cs [hunk 1]
```diff
@@ -503,9 +507,29 @@ private void GenerateEqualLogic(INamedTypeSymbol typeSymbol, IndentedStringBuild
 						}
 						else
 						{
-							using (builder.BlockInvariant($"if(!System.Collections.Generic.EqualityComparer<{type}>.Default.Equals({member.Name}, other.{member.Name}))"))
+							if (equalityMode == KeyEqualityUseEquality)
 							{
-								builder.AppendLineInvariant($"return false; // {member.Name} not equal");
+								using (builder.BlockInvariant($"if(!System.Collections.Generic.EqualityComparer<{type}>.Default.Equals({member.Name}, other.{member.Name}))"))
```
Comment (jeromelaban) [inline src/Uno.CodeGen/EqualityGenerator.cs]: `if(!global::System.Collections`

File: src/Uno.CodeGen/Helpers/TypeSymbolExtensions.cs [hunk 2]
```diff
@@ -153,17 +153,42 @@ bool GetIsImmutable((ITypeSymbol type, bool treatArrayAsImmutable) x)
...
 		}
 
+		private static bool IsImmutableByRequirements(ITypeSymbol type, bool treatArrayAsImmutable)
+		{
+			// Check if type is complying to immutable requirements
+			// 1) all instance fields are readonly
+			// 2) all properties are getter-only
+			// 3) base class is immutable
+
+			if (type
```
Comment (jeromelaban) [inline src/Uno.CodeGen/Helpers/TypeSymbolExtensions.cs]: Looks like the enumeration is done twice and that the filter is very similar, would be interesting to enumerate only once by removing the .OfType<>.

File: src/Uno.CodeGen/EqualityGenerator.cs (hunk 1 above)
Comment (carldebilly) [reply src/Uno.CodeGen/EqualityGenerator.cs]: Done.

File: src/Uno.CodeGen/Helpers/TypeSymbolExtensions.cs (hunk 2 above)
Comment (carldebilly) [reply src/Uno.CodeGen/Helpers/TypeSymbolExtensions.cs]: Done too :-)

**Pull request merged**